lib/src/serializable.dart

```diff
@@ -0,0 +1,237 @@
+    _bridge = null;
+  }
+
+  void registerModule(SerializableModule module) {
+    if (module.serializableKey != null) {
+      _registerForLifecylceEvents(module);
+      _registeredModules[module.serializableKey] = module;
+    } else {
+      _logger.warning('Unable to serialize module without serializableKey');
+    }
+  }
+
+  void deregisterModule(SerializableModule module) {
+    _registeredModules.remove(module.serializableKey);
+  }
+
+  void _registerForAllEvents(SerializableModule module) {
+    if (module.events != null) {
+      for (var event in module.events.allEvents) {
+        if (event is SerializableEvent) {
+          event
+              .listen((payload) => _sendEvent(module, event.eventKey, payload));
+        }
+      }
+    } else {
+      _logger
+          .warning('Events not defined for ${module.serializableKey} module');
+    }
+  }
+
+  void _registerForLifecylceEvents(SerializableModule module) {
+    module.willLoad.listen((_) {
+      _registerForAllEvents(module);
+      _sendEvent(module, 'willLoad', null);
+    });
+    module.didLoad.listen((_) => _sendEvent(module, 'didLoad', null));
+    module.willUnload.listen((_) => _sendEvent(module, 'willUnload', null));
+    module.didUnload.listen((_) {
+      _sendEvent(module, 'didUnload', null);
+      deregisterModule(module);
+    });
+  }
+
+  void _sendEvent(SerializableModule module, String eventKey, Object data) {
+    Map dataToSend = new Map();
+    dataToSend['module'] = module.serializableKey;
+    dataToSend['event'] = eventKey;
+
+    if (data is JsonSerializable) {
+      data = (data as JsonSerializable).toJson();
+    }
+
+    dataToSend['data'] = data;
+
+    _bridge?.broadcast(dataToSend);
+  }
+
+  void _handleApiCall(Map apiCall) {
+    String module = apiCall['module'];
+    String method = apiCall['method'];
+    Object data = apiCall["data"];
+
+    SerializableModule targetModule = _registeredModules[module];
+
+    _deserializeAndCall(targetModule, method, data);
+  }
+
+  void _deserializeAndCall(
+      SerializableModule module, String method, List data) {
+    InstanceMirror apiMirror = reflect(module.api);
+    ClassMirror classMirror = apiMirror.type;
+    MethodMirror apiMethodMirror = classMirror.declarations[new Symbol(method)];
+
+    if (apiMethodMirror == null) {
+      _logger.warning(
+          'Method $method does not exist on ${module.serializableKey}\' module\'s API');
+      return;
+    }
+
+    // Check here that the position args in data match the expected params of the method being called
+    if (data is List && apiMethodMirror.parameters.length == data.length) {
+      for (var i = 0; i < apiMethodMirror.parameters.length; i++) {
+        var param = apiMethodMirror.parameters[i];
+
+        // If the type data being passed to this param is not equal to the expected type
+        // try to serialize it into an Dart class
+        if (param.type.reflectedType != data[i].runtimeType) {
```

Comment:
It will work. After some testing I had to make some changes to how this works. That `if (param.type.reflectedType != data[i].runtimeType` check does not alway work as expected. For example, when passing an array from JS, the type is `JsArray` but if you do a type check on like `jsArray is List` that returns true. So now it checks to see if data being passed is a `Map` and then will try to serialize it into a the expected object of the api method.

I know this is not idea, but before this is used at any type of scale, all this serialization stuff will be replaced by frugal so I think it is ok for now.